### PR TITLE
engine: support deletion

### LINKED ISSUE #73: Roadmap v0.2 - Engine
Provide a simple hash engine that supports `get/set/delete` key-value pairs.
The engine should be able to use different kinds of kernels.

- [x] #168

### LINKED ISSUE #168: Support `delete` operation
#143 only supports `get/set`, we should support `delete` too. This is not as easy as it might sound. We need to add tombstones for deleted records.

### PR DESCRIPTION
**Pull request by huachaohuang**
Closes #168 , closes #73 .

This PR is based on #167 .

### CODE REVIEW

**Comment by huachaohuang** [comment]
This PR is ready now.

**Review by zojw — APPROVED** [review]
LGTM

File: src/engine/hash/src/codec.rs [hunk 1]
```diff
@@ -12,51 +12,102 @@
...
 use crate::{Error, Result};
 
 pub type Timestamp = u64;
 
-pub fn record_size(key: &[u8], value: &[u8]) -> usize {
-    16 + key.len() + value.len()
+#[derive(Clone, Debug, PartialEq)]
+pub enum Value {
+    Put(Vec<u8>),
+    Deletion,
```
Comment (tisonkun) [inline src/engine/hash/src/codec.rs:27]: If you don't extend this enum later, it is exact `Option<Vec<u8>>`. I prepare a pull request https://github.com/huachaohuang/engula/pull/1 that you can make use of.

Comment (huachaohuang) [reply src/engine/hash/src/codec.rs]: Nice work. I think you are very Rusty now.

**Review by tisonkun — APPROVED** [review]

**Pull request merged**